include/GLFW/glfw3.hpp

```diff
@@ -0,0 +1,177 @@
+        versionUnavailable = GLFW_VERSION_UNAVAILABLE,
+        platformError = GLFW_PLATFORM_ERROR,
+        formatUnavailable = GLFW_FORMAT_UNAVAILABLE,
+        noWindowContext = GLFW_NO_WINDOW_CONTEXT,
+        cursorUnavailable = GLFW_CURSOR_UNAVAILABLE,
+        featureUnavailable = GLFW_FEATURE_UNAVAILABLE,
+        featureUnimplemented = GLFW_FEATURE_UNIMPLEMENTED,
+    };
+
+}
+
+namespace std {
+    template <>
+    struct is_error_code_enum<GLFW_HPP_NAMESPACE::ErrorCode> : public true_type {};
+}
+
+namespace GLFW_HPP_NAMESPACE {
+
+    class ErrorCategory : public std::error_category {
+    public:
+        const char *name() const _NOEXCEPT override {
+            return nullptr;
+        }
+
+        std::string message(int __ev) const override {
+            return std::string();
+        }
+    };
+
+    inline std::error_code make_error_code(ErrorCode code) {
+        return {static_cast<int>(code), ErrorCategory()};
+    }
+
+    class Error : public std::system_error {
+    public:
+        ErrorCode const code;
+
+    private:
+        Error(ErrorCode const code, GLFW_HPP_STRING const &description) :
+                std::system_error(std::error_code{code}, description),
+                code{code} {}
+
+        static std::optional<Error> get_() {
+            char const *description;
+            auto code = static_cast<ErrorCode const>(glfwGetError(&description));
+
+            if (code == ErrorCode::noError) {
+                return std::optional<Error>{};
+            }
+
+            return Error(code, description ? _private::makeString_(description) : "");
+        }
+
+        static void getAndThrow_() {
+            auto const err = get_();
+            if (err) {
+                throw err.value();
+            }
+        }
+
+        friend class GLFW;
+        friend class Window;
+    };
+
+    class Window {
+    public:
+        Window(Size const &size, GLFW_HPP_STRING const &title) :
+                size_{size},
+                title_{title},
+                handle_{glfwCreateWindow(size.width, size.height, _private::cString_(title), nullptr, nullptr)} {
+            Error::getAndThrow_();
+        }
+
+        void makeContextCurrent() const {
+            glfwMakeContextCurrent(this->handle_);
+        }
+
+        [[nodiscard]] bool shouldClose() const {
+            return glfwWindowShouldClose(this->handle_);
+        }
+
+        void swapBuffers() const {
+            glfwSwapBuffers(this->handle_);
+        }
+
+        GLFWwindow *handle_;
+    private:
+
+        Size size_;
+        GLFW_HPP_STRING title_;
+    };
+
+    class GLFW {
+    public:
+        GLFW() {
+            if (!glfwInit()) {
+                Error::getAndThrow_();
+            }
+        }
+
+        explicit GLFW(ErrorFun errorCallback) : GLFW() {
+            this->errorCallback = errorCallback;
+            glfwSetErrorCallback(this->errorCallback);
+        }
+
+        virtual ~GLFW() {
```

Comment:
Very small nitpick: the `GLFW` type does not seem like a polymorphic type, but has a virtual destructor. Unless you explicitly intend users to extend and use the `GLFW` type in a polymorphic context, I would remove the `virtual` there.

Reply:
Perhaps the `final` specifier could be also added to prevent the class from being inherited from.